File: horoscope.py
#!/usr/bin/env python3

import os
import sys
import shutil

def show_help(script_name):
  """Displays help information."""
  print(f"Usage: {script_name} [options] <output_filename>")
  print("Gather project files and metadata for analysis.")
  print("Options:")
  print("  -h, --help        Display this help message.")
  print("  -n, --name        Name the output file (default: project_files.txt).")
  print("  -f, --firebase     Include Firebase config files.")
  print("  -d, --dart        Analyze a Dart project instead of Python.")
  print("  -i, --install     Install this script to /usr/local/bin (requires sudo).")
  print("")
  print("Run this from your project's root directory.")

def main():
  """The main function to gather project files."""
  output_filename = "project_files.txt"
  include_firebase = False
  include_dart = False  # Now defaults to Python

  # Parse command-line options.
  i = 1
  while i < len(sys.argv):
    arg = sys.argv[i]
    if arg in ('-h', '--help'):
      show_help(sys.argv[0])
      sys.exit(0)
    elif arg in ('-n', '--name'):
      if i + 1 < len(sys.argv):
        output_filename = sys.argv[i + 1]
        i += 1
      else:
        print("Error: Missing filename after -n option.")
        show_help(sys.argv[0])
        sys.exit(1)
    elif arg in ('-f', '--firebase'):
      include_firebase = True
    elif arg in ('-d', '--dart'):  # Changed to -d for Dart
      include_dart = True
    elif arg in ('-i', '--install'):
      if os.geteuid() != 0:
        print('Error: Installation requires root privileges. Use sudo.')
        sys.exit(1)

      install_path = '/usr/local/bin'
      home_dir = os.environ.get('HOME')
      if home_dir and os.path.isdir(os.path.join(home_dir, '.local/bin')):
        install_path = os.path.join(home_dir, '.local/bin')
      script_path = os.path.join(install_path, 'gather_project_files')  # More descriptive name

      try:
        shutil.copy(sys.argv[0], script_path)
        os.chmod(script_path, 0o755)
        print(f'Script installed successfully to {script_path}')
      except Exception as e:
        print(f'Error during installation: {e}')
        sys.exit(1)
      sys.exit(0)
    else:
      print(f"Error: Invalid option: {arg}")
      show_help(sys.argv[0])
      sys.exit(1)
    i += 1

  os.makedirs(os.path.dirname(output_filename), exist_ok=True)

  with open(output_filename, 'w') as output_file:
    pass

  if include_dart:  # Logic adjusted for Dart
    try:
      with open('pubspec.yaml', 'r') as pubspec_file:
        pubspec_content = pubspec_file.read()
        project_name = next((line.split(':')[1].strip() for line in pubspec_content.splitlines() if line.startswith('name:')), None)
        project_description = next((line.split(':')[1].strip() for line in pubspec_content.splitlines() if line.startswith('description:')), None)
        output_file.write(f'Project Name: {project_name}\n')
        output_file.write(f'Description: {project_description}\n')
        output_file.write('---- pubspec.yaml ----\n')
        output_file.write(pubspec_content)
    except FileNotFoundError:
      print("Error: Cannot find pubspec.yaml. Are you in a Dart project directory?")
      sys.exit(1)
  else:
    # Check for Python "metadata" files
    for file_name in ('requirements.txt', 'setup.py', 'pyproject.toml'):
      try:
        with open(file_name, 'r') as f:
          with open(output_filename, 'a') as output_file:
            output_file.write(f'---- {file_name} ----\n')
            output_file.write(f.read())
      except FileNotFoundError:
        pass

  for root, _, files in os.walk('.'):
    for file in files:
      file_path = os.path.join(root, file)
      if (not include_dart and file_path.endswith('.py') or  # Python first
          include_dart and file_path.endswith('.dart') or
          include_dart and not '.' in file or
          include_firebase and ('firebase.json' in file_path or '.firebase.' in file_path)):
        with open(output_filename, 'a') as output_file:
          output_file.write(f'---- {file_path} ----\n')
          with open(file_path, 'r') as f:
            output_file.write(f.read())

  print(f'Output written to: {output_filename}')

if __name__ == "__main__":
  main()

File: horoscope.dart
#!/usr/bin/env dart

// Importing the necessary libraries for file and directory operations.
import 'dart:io';

// Function to display help information.
void showHelp(String scriptName) {
  print("Usage: $scriptName [options] <output_filename>");
  print("Gather project files and metadata for analysis.");
  print("Options:");
  print("  -h, --help         Display this help message.");
  print("  -n, --name         Name the output file (default: flutter_dart_files.txt).");
  print("  -f, --firebase     Include Firebase config files (if used).");
  print("  -p, --pythonaconda  Analyze a Python project instead of Dart.");
  print("  -i, --install      Install horoscope to /usr/local/bin (requires sudo).");
  print("");
  print("Run this from your project's root directory.");
}

void main(List<String> arguments) {
  // Default settings
  var outputFilename = "flutter_dart_files.txt";
  var includeFirebase = false;
  var includePython = false;

  // Parse command-line options.
  for (var i = 0; i < arguments.length; i++) {
    switch (arguments[i]) {
      case '-h':
      case '--help':
        showHelp(arguments[0]);
        exit(0);
      case '-n':
      case '--name':
        if (i + 1 < arguments.length) {
          outputFilename = arguments[i + 1];
          i++;
        } else {
          print("Error: Missing filename after -n option.");
          showHelp(arguments[0]);
          exit(1);
        }
        break;
      case '-f':
      case '--firebase':
        includeFirebase = true;
        break;
      case '-p':
      case '--pythonaconda':
        includePython = true;
        break;
      case '-i':
      case '--install':
        // Check for sudo privileges
        if (Platform.environment['USER'] != 'root') {
          print('Error: Installation requires root privileges. Use sudo.');
          exit(1);
        }

        // Determine the appropriate installation path
        var installPath = '/usr/local/bin'; // Default path
        var homeDir = Platform.environment['HOME'];
        if (homeDir != null && Directory('$homeDir/.local/bin').existsSync()) {
          installPath = '$homeDir/.local/bin'; // Prioritize user's local bin
        }
        var scriptPath = '$installPath/horoscope';

        try {
          File(Platform.script.toFilePath()).copySync(scriptPath);
          Process.runSync('chmod', ['+x', scriptPath]);
          print('Horoscope installed successfully to $scriptPath');
        } catch (e) {
          print('Error during installation: $e');
          exit(1);
        }
        exit(0); // Exit after installation
      default:
        print("Error: Invalid option: ${arguments[i]}");
        showHelp(arguments[0]);
        exit(1);
    }
  }

  // Create the output directory if it doesn't exist.
  var outputDir = Directory(outputFilename).parent;
  if (!outputDir.existsSync()) {
    outputDir.createSync(recursive: true);
  }

  // Clear the output file.
  var outputFile = File(outputFilename);
  outputFile.writeAsStringSync('');

  // Get project metadata (pubspec.yaml for Dart, or Python equivalents)
  if (!includePython) {
    var pubspecFile = File('pubspec.yaml');
    if (pubspecFile.existsSync()) {
      var pubspecContent = pubspecFile.readAsStringSync();
      var projectName = RegExp(r'name:\s*([^\n]+)').firstMatch(pubspecContent)?.group(1)?.trim();
      var projectDescription = RegExp(r'description:\s*([^\n]+)').firstMatch(pubspecContent)?.group(1)?.trim();
      outputFile.writeAsStringSync('Project Name: $projectName\n', mode: FileMode.append);
      outputFile.writeAsStringSync('Description: $projectDescription\n', mode: FileMode.append);
      outputFile.writeAsStringSync('---- pubspec.yaml ----\n', mode: FileMode.append);
      outputFile.writeAsStringSync(pubspecContent, mode: FileMode.append);
    } else {
      print("Error: Cannot find pubspec.yaml. Are you in the right directory?");
      exit(1);
    }
  } else {
    // Check for Python "metadata" files
    var requirementsFile = File('requirements.txt');
    var setupFile = File('setup.py');
    var pyprojectFile = File('pyproject.toml');

    if (requirementsFile.existsSync()) {
      outputFile.writeAsStringSync('---- requirements.txt ----\n', mode: FileMode.append);
      outputFile.writeAsStringSync(requirementsFile.readAsStringSync(), mode: FileMode.append);
    }
    if (setupFile.existsSync()) {
      outputFile.writeAsStringSync('---- setup.py ----\n', mode: FileMode.append);
      outputFile.writeAsStringSync(setupFile.readAsStringSync(), mode: FileMode.append);
    }
    if (pyprojectFile.existsSync()) {
      outputFile.writeAsStringSync('---- pyproject.toml ----\n', mode: FileMode.append);
      outputFile.writeAsStringSync(pyprojectFile.readAsStringSync(), mode: FileMode.append);
    }
  }

  // Gather ALL files in ONE glorious loop!
  var files = Directory.current.listSync(recursive: true);
  for (var file in files) {
    if (file is File &&
      (includePython && file.path.endsWith('.py') ||
      !includePython && file.path.endsWith('.dart') ||
      (!includePython && !file.path.contains('.')) ||  // Grab files without extensions
      (includeFirebase && (file.path.endsWith('firebase.json') || file.path.contains('.firebase.'))))) {
      outputFile.writeAsStringSync('---- ${file.path} ----\n', mode: FileMode.append);
    outputFile.writeAsStringSync(file.readAsStringSync(), mode: FileMode.append);
      }
  }

  // Notify the user.
  print('Output written to: $outputFilename');
}
